src/threading_utils/parallel_for.h

```diff
@@ -0,0 +1,83 @@
+/*!
+* Copyright (c) 2021 by Contributors
+* \file parallel_for.h
+* \brief Implemenation of parallel for loop
+* \author Hyunsu Cho
+*/
+#ifndef TREELITE_THREADING_UTILS_PARALLEL_FOR_H_
+#define TREELITE_THREADING_UTILS_PARALLEL_FOR_H_
+
+#include <treelite/logging.h>
+#include <thread>
+#include <algorithm>
+#include <vector>
+#include <cstddef>
+
+namespace treelite {
+namespace threading_utils {
+
+template <typename IndexType>
+std::vector<IndexType> ComputeWorkRange(IndexType begin, IndexType end, std::size_t nthread);
+
+template <typename IndexType, typename FuncType>
+void ParallelFor(IndexType begin, IndexType end, std::size_t nthread, FuncType func) {
+  TREELITE_CHECK_GT(nthread, 0) << "nthread must be positive";
+  TREELITE_CHECK_LE(nthread, std::thread::hardware_concurrency())
+    << "nthread cannot exceed " << std::thread::hardware_concurrency();
+  if (begin == end) {
+    return;
+  }
+  /* Divide the rnage [begin, end) equally among the threads.
+   * The i-th thread gets the range [work_range[i], work_range[i+1]). */
+  std::vector<IndexType> work_range = ComputeWorkRange(begin, end, nthread);
+
+  // Launch (nthread - 1) threads, as the main thread should also perform work.
+  std::vector<std::thread> threads;
+  for (std::size_t thread_id = 1; thread_id < nthread; ++thread_id) {
+    threads.emplace_back([&work_range, &func, thread_id]() {
```

Comment:
Do we need to worry about exception handling?  If so then you might want to use async + future.  Otherwise looks good.

Reply:
Good point. Let me change it to `std::async` instead.